pkg/metrics/metrics.go

```diff
@@ -0,0 +1,176 @@
+		viewsToUnregister = append(viewsToUnregister, runsNotStoredView)
+	}
+	if runStorageLatencyView != nil {
+		viewsToUnregister = append(viewsToUnregister, runStorageLatencyView)
+	}
+	view.Unregister(viewsToUnregister...)
+	registeredAt = nil
+}
+
+// IdempotentRegisterViews Ensures all shared views are registered.
+// Does not unregister views if they have already been registered.
+func IdempotentRegisterViews(logger *zap.SugaredLogger) {
+	registerMutex.Lock()
+	defer registerMutex.Unlock()
+	if registeredAt != nil {
+		return
+	}
+	unregisterViews(logger)
+	if err := registerViews(logger); err != nil {
+		logger.Errorf("Failed to register View %v ", err)
+	} else {
+		now := time.Now()
+		registeredAt = &now
+	}
+}
+
+// CountRunNotStored records a run that was not stored due to deletion or timeout
+func CountRunNotStored(ctx context.Context, namespace, kind string) error {
+	ctx, err := tag.New(
+		ctx,
+		tag.Insert(kindTag, kind),
+		tag.Insert(namespaceTag, namespace),
+	)
+	if err != nil {
+		return fmt.Errorf("unable to create tags for %s metric: %w", runsNotStoredCount.Name(), err)
+	}
+
+	metrics.Record(ctx, runsNotStoredCount.M(1))
+	return nil
+}
+
+// OnStore returns a function that checks if metrics are configured for a config.Store, and registers it if so
+func OnStore(logger *zap.SugaredLogger) func(name string, value any) {
+	return func(name string, value any) {
+		if name != config.GetMetricsConfigName() {
+			return
+		}
+		_, ok := value.(*config.Metrics)
+		if !ok {
+			logger.Error("Failed to do type insertion for extracting metrics config")
+			return
+		}
+		// For shared metrics, we use idempotent registration
+		IdempotentRegisterViews(logger)
+	}
+}
+
+// RecordStorageLatency records the storage latency metric for both PipelineRuns and TaskRuns
+func (r *Recorder) RecordStorageLatency(ctx context.Context, object interface{}) error {
+	var (
+		completionTime *time.Time
+		namespace      string
+		kind           string
+	)
+
+	// Extract completion time and metadata using type switch
+	switch o := object.(type) {
+	case *pipelinev1.PipelineRun:
+		if o.Status.CompletionTime == nil {
+			return nil
+		}
+		completionTime = &o.Status.CompletionTime.Time
+		namespace = o.Namespace
+		kind = "pipelinerun"
+
+	case *pipelinev1.TaskRun:
+		if o.Status.CompletionTime == nil {
+			return nil
+		}
+		completionTime = &o.Status.CompletionTime.Time
+		namespace = o.Namespace
+		kind = "taskrun"
+
+	default:
+		return fmt.Errorf("unsupported object type: %T", object)
+	}
+
+	// Calculate latency from completion to now
+	now := r.clock.Now()
+	latency := now.Sub(*completionTime)
+
+	// Create tags
+	tags := []tag.Mutator{
+		tag.Insert(kindTag, kind),
+		tag.Insert(namespaceTag, namespace),
+	}
+
+	ctx, err := tag.New(ctx, tags...)
+	if err != nil {
+		return fmt.Errorf("error creating tagged context: %w", err)
+	}
+
+	// Record the metric
+	metrics.Record(ctx, runStorageLatency.M(float64(latency/time.Second)))
```

Comment:
This call could happen multiple times, if we disable incomplete runs isn't set.

Reply:
I don't think that's the case. 
We are calling this function when stored annotation is present and I can see that stored annotation is only applied when run is marked as completed. Here is the code https://github.com/tektoncd/results/blob/main/pkg/watcher/reconciler/dynamic/dynamic.go#L746